llvm/include/llvm/Analysis/TargetTransformInfoImpl.h

```diff
@@ -975,9 +975,10 @@ class TargetTransformInfoImplBase {
     return 0;
   }
 
-  virtual bool getTgtMemIntrinsic(IntrinsicInst *Inst,
-                                  MemIntrinsicInfo &Info) const {
-    return false;
+  virtual std::pair<std::optional<MemIntrinsicInfo>,
+                    std::optional<SmallVector<InterestingMemoryOperand, 1>>>
```

Comment:
Do we need an optional around the SmallVector? Can we use an empty vector to represent no interesting operand?

Looks like EarlyCSE will ignore MemIntrinsicInfo with a null `PtrVal`. Can we use that instead of an optional?

Reply:
I think use an empty vector to represent no interesting operand is a great idea, in this way we don't need to use optional for SmallVector.

Use a null `PtrVal` to check MemIntrinsicInfo might be fine, I just checked the target that uses getTgtMemIntrinsic and find that all of them fill `PtrVal` and EarlyCSE/LoopStrengthReduce also ignore `PtrVal` with nullptr.

I will try to modify it with dropping optional.

Reply:
```suggestion
  std::pair<MemIntrinsicInfo,
            SmallVector<InterestingMemoryOperand, 1>>
```
The interface after dropping `std::optional` will look like this.